### DISCUSSION TITLE
Notification error with mattermost

### DISCUSSION THREAD

**Discussion by larrieu-olivier** [PrefectHQ/prefect#8141] [Q&A] [answered]
i'm trying to use the notification system to notify to a mattermost channel but that is not working.
To do this I use the slack notification configuration in the UI, I put my channel mattermost notification URL.
I start a flow and then it end successfully and  prefect try to send notification but emmit log errors.

```text
14:34:49.422 | ERROR   | apprise - Unparseable URL https://m...o/hooks/w...r
14:34:49.424 | ERROR   | apprise - There are no service(s) to notify
```
I would like to use the UI configurator instead of code based notification.
is there someone which configured mattermost notifications ?

regards

**Comment by rpeden** [comment, marked as answer]
Apprise should be able to parse the URL if you change `https://` to `mmosts://` and remove `hooks`, because Apprise automatically inserts it.

Unless you are running your own Mattermost server on port 8065, you'll need to add the port as well. So, for example, `https://your-mattermost-host/hooks/abcdefg` would become `mmosts://your-mattermost-host:443/abcdefg` assuming you are connecting to a Mattermost server using the standard SSL/TLS port 443. 

I just did a test run using this URL format with a Mattermost Cloud channel, and a Prefect notification appeared as expected. 😃

**Reply by larrieu-olivier** [reply]
thank you, i ll give a try.
regards

**Comment by Pheggas** [comment]
Hello, @rpeden .
I'm getting similar issue. But with the new Prefect 3, the block is now defined in UI and not by editing the file. I'm trying to push notifications to https Mattermost but getting error:

```
Oct 24 12:41:02 redacted-prefect prefect[562269]: 12:41:02.268 | INFO    | apprise - Notifying 1 service(s) asynchronously.
Oct 24 12:41:02 redacted-prefect prefect[562269]: 12:41:02.271 | DEBUG   | apprise - Mattermost POST URL: http://mattermost.example.com:443/hooks/redacted (cert_verify=True)
Oct 24 12:41:02 redacted-prefect prefect[562269]: 12:41:02.272 | DEBUG   | apprise - Mattermost Payload: {'text': "Prefect work pool 'Redacted' has entered status 'WorkPoolStatus.NOT_READY'\r\nName: Redacted\nStatus: WorkPoolStatus.NOT_READY\nURL: https://0.0.0.0:4200/work-pools/work-pool/Redacted", 'icon_url': None, 'username': 'Prefect Notifications', 'c>
Oct 24 12:41:02 redacted-prefect prefect[562269]: Invalid HTTP request received.
Oct 24 12:41:02 redacted-prefect prefect[562269]: 12:41:02.336 | ERROR   | GlobalEventLoopThread | prefect._internal.concurrency - Service 'EventsWorker' failed with 1 pending items.
Oct 24 12:41:02 redacted-prefect prefect[562269]: 12:41:02.347 | WARNING | apprise - Failed to send Mattermost notification to channel test-prefect-notifications: Bad Request - Unsupported Parameters., error=400.
Oct 24 12:41:02 redacted-prefect prefect[562269]: 12:41:02.348 | DEBUG   | apprise - Response Details:
Oct 24 12:41:02 redacted-prefect prefect[562269]: b'<html>\r\n<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<center>The plain HTTP request was sent to HTTPS port</center>\r\n<hr><center>nginx/1.18.0 (Ubuntu)</center>\r\n</body>\r\n</html>\r\n' 
```

Which is okay as it automatically puts http in front of the DNS name.

Block:
![Snímka obrazovky 2024-10-24 o 14 55 44](https://github.com/user-attachments/assets/c60a5af4-7bc5-4432-b5e7-cd8f719d137b)

**Reply by Pheggas** [reply]
I was able to resolve this. For anyone having this issue, please, install slack webhook integration and paste the webhook there in form like: `mmosts://mattermost.example.com:443/x3i7w6mdgh6vgfqb0vdp4e3fj0h` (domain and hook token redacted). Turned out that there **MUSN'T**  be a `/hooks/` path.

**Reply by hermonjaycatalyst** [reply]
I am experiencing the same issue, however, the slack webhook trick doesn't work. How did you to push notifications to https Mattermost? I also want to see the response of the Mattermost block. @Pheggas

**Reply by Pheggas** [reply]
Just like i mentioned above. No changed made. In case you are interested in what the prefect is actually pushing to mattermost api, the sample json body looks like this:
```
{
  "version": "1.0",
  "title": "Prefect work pool 'Test-pool' has entered status 'WorkPoolStatus.PAUSED'",
  "message": "Name: Test-pool\nStatus: WorkPoolStatus.PAUSED\nURL: https://prefect.internal.example.com/work-pools/work-pool/Test-pool",
  "attachments": [],
  "type": "prefect_default"
}
```

or crash:

```
{
  "version": "1.0",
  "title": "Prefect flow run alert",
  "message": "Flow run update-proxies-flow/sticky-chachalaca observed in state `Crashed` at 2025-02-18 15:14:54.076554+00:00.\nFlow ID: 94e5f44e-14ca-4652-8c68-26993ae559a1\nFlow run ID: d1e3b7f2-4a89-4c56-9d3a-5f8b2e7c1a4f\nFlow run URL: None\nWorkpool: Test-pool\nState message: Flow run could not be submitted to infrastructure",
  "attachments": [],
  "type": "prefect_default"
}
```

**Comment by hnykda** [comment]
OK, I found my way. First of all, you definitely need the `webhook` (in Mattermost console: Integrations -> Incoming Webhook), not `bot`. It's misleading because Prefect UI asks you for token, and token is present with the bot but explicitly not with webhook. Anyway, generate a webhook, it will give you the following URL format (I assume you are running behind HTTPS, which you probably do unless you live under a rock or something, i.e. for me it means that mattermost is accessible under https://<your-url> ):

```
URL: https://<your-url>/hooks/uh3h<some-gibberish>gsua
```

and this is how you torture it into Prefect UI format:
![image](https://github.com/user-attachments/assets/ef7a5b80-4263-4fe4-85d2-715474624aa9)
